### ISSUE TITLE
Native age identity support

### ISSUE THREAD

**Issue by str4d** [str4d/age-plugin-yubikey#174]
[YubiKeys with firmware 5.7.0 and above have support for X25519](https://github.com/Yubico/yubico-piv-tool/blob/master/debian/changelog) (thanks https://github.com/smlx/piv-agent/issues/134 for making me aware of this).

It should be possible to generate native age identities in PIV slots, and use them just like P-256 identities on the decryption side (our identity format points to a specific YubiKey serial and slot, from which we can then determine whether or not the key is P-256 or X25519). On the encryption side, these would just be regular native age recipients (and thus not require a plugin to encrypt to).

The downside is that because native age identities do not include any sort of tag, we would need to assume that any X25519 YubiKey slot could potentially match, which means in multi-key settings the UX is not as great, but that is a trade-off we can probably explain in the setup TUI.

**Comment by pinpox** [comment]
Would this allow for using existing age (or ssh) keys on the yubikey? I"m searching for a solution that let's me use the yubikey for age encryption/decryption while having a offline (printed out) backup of my key so that I can restore it on a new yubikey in case of I lose it.

**Comment by dlubawy** [comment]
This is definitely possible. I created a [PR over at yubikey.rs](https://github.com/iqlusioninc/yubikey.rs/pull/577) to add curve 25519 support to the underlying yubikey crate. I then forked [this repo on a new branch](https://github.com/dlubawy/age-plugin-yubikey/tree/combined_encryption) to test it all out together. It seems to work alright (though the code isn't great since I'm still new to Rust).

Some things to consider:
* X25519 keys cannot sign certs so I had the plugin export an attestation cert and import to the key's slot on build instead
  - This additionally means that X25519 keys cannot be named using `TAG_HEX`
  - Any X25519 key with an attestation cert from YubiKey is considered a valid age identity for the plugin
* Native age keys are supported through this, but it requires some effort since they cannot sign themselves. Steps are:
  - Decode the bech32 of an age identity to get the X25519 private key
  - Generate a self-signed certificate (using another key capable of signing) for the corresponding public key (Organization set to 'age-plugin-yubikey' for it to show in the `--list`)
  - Import the key/cert pair to slot of choice

**Comment by pinpox** [comment]
@str4d Would you be interested/have the time to implement this? If it's just having a new yubi that's missing for this, I would offer to sponsor one.

**Comment by str4d** [comment]
I've now obtained a 5.7 firmware YubiKey, so am going to see if I can get this into the next release.

**Comment by pinpox** [comment]
> I've now obtained a 5.7 firmware YubiKey, so am going to see if I can get this into the next release.

Very nice to hear, thanks for working on this, I'm still very much interest in this feature.

**Comment by dlubawy** [comment]
Update on this. I rebased [my PR over at yubikey.rs](https://github.com/iqlusioninc/yubikey.rs/pull/577) (really a new branch with the ref changed to point to a new head that is able to be merged with main). My yubikeys all have their management keys and pins changed from defaults so I couldn't use the regular integration tests there. Decided to pickup the branch @str4d started for the update to the latest version of the lib and [ported my old code over onto a new branch using that as a base](https://github.com/dlubawy/age-plugin-yubikey/tree/rebase/combined_encryption) to test the lib with. Seems to work, so I think this should all still be doable so long as the yubikey.rs code can be merged and this project can update to its latest version.

**Comment by quite** [comment]
> The downside is that because native age identities do not include any sort of tag, we would need to assume that any X25519 YubiKey slot could potentially match, which means in multi-key settings the UX is not as great, but that is a trade-off we can probably explain in the setup TUI.

Yeah I ran into something similar when building [age-plugin-tkey](https://github.com/quite/age-plugin-tkey). The [X25519 device app](https://github.com/quite/tkey-device-x25519) that it runs on the TKey also uses true X25519 keys, so the recipient looks the same as one generated by `age-keygen`. So when decrypting multiple recipients, each recipient has to be tried on the TKey (in the current impl., that also forces the user to touch the TKey on every attempt -- unless touch requirement has been disabled for the identity, but that's not the default).

Not great for UX, as written about here: https://github.com/quite/age-plugin-tkey?tab=readme-ov-file#multiple-recipients but feels like it's the right way anyhow. For age-plugin-tkey, I have [ideas](https://github.com/quite/age-plugin-tkey/issues/4) for an alternative device app impl, which could alleviate this.
